File: packages/backend/src/services/stockService.ts
import axios from 'axios'
import iconv from 'iconv-lite'

export interface StockQuote {
  code: string
  name: string
  price: number
  change: number
  changePercent: number
  volume: number
  amount: number
  high: number
  low: number
  open: number
  prevClose: number
  market: string
  updateTime: string
}

export interface StockDetail extends StockQuote {
  // 买卖盘
  bidPrice: number[]
  bidVolume: number[]
  askPrice: number[]
  askVolume: number[]
  // 基本面
  turnoverRate: number
  peRatio: number
  pbRatio: number
  totalMarketCap: number
  floatMarketCap: number
  totalShares: number
  floatShares: number
  // 价格统计
  amplitude: number
  volumeRatio: number
  commissionRatio: number
  avgPrice: number
  week52High: number
  week52Low: number
  // 近期涨跌
  change5d: number
  change10d: number
  change20d: number
  // K线数据（日K）
  kline: KLineData[]
}

export interface KLineData {
  date: string
  open: number
  close: number
  high: number
  low: number
  volume: number
  amount: number
}

export interface CommodityQuote {
  code: string
  name: string
  price: number
  changePercent: number
  open: number
  prevClose: number
  high: number
  low: number
  updateTime: string
}

const TENCENT_API = 'https://qt.gtimg.cn/q='

const STOCK_LIST: Array<{ code: string; name: string; market: string }> = [
  { code: 'sh600519', name: '贵州茅台', market: '上海' },
  { code: 'sz000001', name: '平安银行', market: '深圳' },
  { code: 'sz000002', name: '万科A', market: '深圳' },
  { code: 'sz300033', name: '同花顺', market: '深圳' },
  { code: 'sz000858', name: '五粮液', market: '深圳' },
  { code: 'sz002594', name: '比亚迪', market: '深圳' },
  { code: 'sh600036', name: '招商银行', market: '上海' },
  { code: 'sh601318', name: '中国平安', market: '上海' },
  { code: 'sh600276', name: '恒瑞医药', market: '上海' },
  { code: 'sz000725', name: '京东方A', market: '深圳' },
  { code: 'sh600900', name: '长江电力', market: '上海' },
  { code: 'sz002415', name: '海康威视', market: '深圳' },
]

const COMMODITY_LIST: Array<{ code: string; name: string }> = [
  { code: 'hf_GC', name: 'COMEX黄金' },
  { code: 'hf_CL', name: 'NYMEX原油' },
  { code: 'hf_HG', name: 'COMEX铜' },
  { code: 'hf_SI', name: 'COMEX白银' },
  { code: 'hf_NG', name: 'NYMEX天然气' },
]

const INDEX_CODES = [
  { code: 'sh000001', name: '上证指数' },
  { code: 'sz399001', name: '深证成指' },
  { code: 'sz399006', name: '创业板指' },
  { code: 'sh000688', name: '科创50' },
]

const cache = new Map<string, { data: unknown; timestamp: number }>()
const CACHE_TTL = 5000

async function fetchTencentGBK(url: string): Promise<string> {
  const response = await axios.get(url, {
    timeout: 10000,
    responseType: 'arraybuffer',
    headers: { Referer: 'https://stock.finance.qq.com' },
  })
  return iconv.decode(Buffer.from(response.data), 'gbk')
}

export function parseTencentResponse(raw: string): StockQuote[] {
  const results: StockQuote[] = []
  const lines = raw.trim().split(';').filter(Boolean)

  for (const line of lines) {
    const match = line.match(/v_[^=]+="(.+)"/)
    if (!match) continue

    const parts = match[1].split('~')
    if (parts.length < 45) continue

    const name = parts[1]
    const code = parts[2]
    const price = parseFloat(parts[3]) || 0
    const prevClose = parseFloat(parts[4]) || 0
    const open = parseFloat(parts[5]) || 0
    const volume = parseInt(parts[6]) || 0
    const high = parseFloat(parts[33]) || 0
    const low = parseFloat(parts[34]) || 0
    const amount = parseFloat(parts[37]) || 0
    const updateTime = parts[45] || ''

    const change = price - prevClose
    const changePercent = prevClose > 0 ? (change / prevClose) * 100 : 0

    const stockInfo = STOCK_LIST.find((s) => s.code === `sh${code}` || s.code === `sz${code}`)

    results.push({
      code,
      name: name || stockInfo?.name || code,
      price,
      change: Math.round(change * 100) / 100,
      changePercent: Math.round(changePercent * 100) / 100,
      volume,
      amount,
      high,
      low,
      open,
      prevClose,
      market: stockInfo?.market || (code.startsWith('6') ? '上海' : '深圳'),
      updateTime,
    })
  }

  return results
}

// 解析单只股票详情（扩展字段）
export function parseStockDetail(raw: string): StockDetail | null {
  const lines = raw.trim().split(';').filter(Boolean)
  if (lines.length === 0) return null

  const match = lines[0].match(/v_[^=]+="(.+)"/)
  if (!match) return null

  const p = match[1].split('~')
  if (p.length < 50) return null

  const code = p[2]
  const price = parseFloat(p[3]) || 0
  const prevClose = parseFloat(p[4]) || 0
  const open = parseFloat(p[5]) || 0
  const volume = parseInt(p[6]) || 0
  const high = parseFloat(p[33]) || 0
  const low = parseFloat(p[34]) || 0
  const amount = parseFloat(p[37]) || 0

  // 买卖盘 10档
  const bidPrice: number[] = []
  const bidVolume: number[] = []
  const askPrice: number[] = []
  const askVolume: number[] = []
  for (let i = 0; i < 5; i++) {
    bidPrice.push(parseFloat(p[9 + i * 2]) || 0)
    bidVolume.push(parseInt(p[10 + i * 2]) || 0)
    askPrice.push(parseFloat(p[19 + i * 2]) || 0)
    askVolume.push(parseInt(p[20 + i * 2]) || 0)
  }

  const change = price - prevClose
  const changePercent = prevClose > 0 ? (change / prevClose) * 100 : 0
  const stockInfo = STOCK_LIST.find((s) => s.code === `sh${code}` || s.code === `sz${code}`)

  return {
    code,
    name: p[1] || stockInfo?.name || code,
    price,
    change: Math.round(change * 100) / 100,
    changePercent: Math.round(changePercent * 100) / 100,
    volume,
    amount,
    high,
    low,
    open,
    prevClose,
    market: stockInfo?.market || (code.startsWith('6') ? '上海' : '深圳'),
    updateTime: p[30] || '',
    bidPrice,
    bidVolume,
    askPrice,
    askVolume,
    turnoverRate: parseFloat(p[38]) || 0,
    peRatio: parseFloat(p[39]) || 0,
    pbRatio: parseFloat(p[46]) || 0,
    totalMarketCap: parseFloat(p[44]) || 0,
    floatMarketCap: parseFloat(p[45]) || 0,
    totalShares: parseFloat(p[67]) || 0,
    floatShares: parseFloat(p[68]) || 0,
    amplitude: parseFloat(p[43]) || 0,
    volumeRatio: parseFloat(p[49]) || 0,
    commissionRatio: parseFloat(p[50]) || 0,
    avgPrice: parseFloat(p[51]) || 0,
    week52High: parseFloat(p[47]) || 0,
    week52Low: parseFloat(p[48]) || 0,
    change5d: parseFloat(p[62]) || 0,
    change10d: parseFloat(p[63]) || 0,
    change20d: parseFloat(p[64]) || 0,
    kline: [],
  }
}

export function parseCommodityResponse(raw: string): CommodityQuote[] {
  const results: CommodityQuote[] = []
  const lines = raw.trim().split(';').filter(Boolean)

  for (const line of lines) {
    const match = line.match(/v_([^=]+)="(.+)"/)
    if (!match) continue

    const code = match[1]
    const parts = match[2].split(',')
    if (parts.length < 8) continue

    const price = parseFloat(parts[0]) || 0
    const changePercent = parseFloat(parts[1]) || 0
    const open = parseFloat(parts[2]) || 0
    const prevClose = parseFloat(parts[3]) || 0
    const high = parseFloat(parts[4]) || 0
    const low = parseFloat(parts[5]) || 0
    const updateTime = parts[6] || ''

    const info = COMMODITY_LIST.find((c) => c.code === code)

    results.push({
      code,
      name: info?.name || code,
      price: Math.round(price * 100) / 100,
      changePercent: Math.round(changePercent * 100) / 100,
      open: Math.round(open * 100) / 100,
      prevClose: Math.round(prevClose * 100) / 100,
      high: Math.round(high * 100) / 100,
      low: Math.round(low * 100) / 100,
      updateTime,
    })
  }

  return results
}

// 解析腾讯K线数据
export function parseKLineResponse(raw: string, stockCode: string): KLineData[] {
  try {
    const data = JSON.parse(raw)
    const lines = data?.data?.[stockCode]?.day || []
    return lines.map((line: string[]) => ({
      date: line[0],
      open: parseFloat(line[1]) || 0,
      close: parseFloat(line[2]) || 0,
      high: parseFloat(line[3]) || 0,
      low: parseFloat(line[4]) || 0,
      volume: parseInt(line[5]) || 0,
      amount: parseFloat(line[6]) || 0,
    }))
  } catch {
    return []
  }
}

export async function getRealTimeQuotes(codes?: string[]): Promise<StockQuote[]> {
  const targetCodes = codes && codes.length > 0 ? codes : STOCK_LIST.map((s) => s.code)
  const cacheKey = targetCodes.join(',')

  const cached = cache.get(cacheKey)
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.data as StockQuote[]
  }

  const url = `${TENCENT_API}${targetCodes.join(',')}`
  const raw = await fetchTencentGBK(url)
  const data = parseTencentResponse(raw)
  cache.set(cacheKey, { data, timestamp: Date.now() })
  return data
}

export async function getMarketIndices(): Promise<StockQuote[]> {
  const cacheKey = 'indices'
  const cached = cache.get(cacheKey)
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.data as StockQuote[]
  }

  const url = `${TENCENT_API}${INDEX_CODES.map((i) => i.code).join(',')}`
  const raw = await fetchTencentGBK(url)
  const data = parseTencentResponse(raw)
  cache.set(cacheKey, { data, timestamp: Date.now() })
  return data
}

export async function getCommodities(): Promise<CommodityQuote[]> {
  const cacheKey = 'commodities'
  const cached = cache.get(cacheKey)
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.data as CommodityQuote[]
  }

  const url = `${TENCENT_API}${COMMODITY_LIST.map((c) => c.code).join(',')}`
  const raw = await fetchTencentGBK(url)
  const data = parseCommodityResponse(raw)
  cache.set(cacheKey, { data, timestamp: Date.now() })
  return data
}

// 获取单只股票详情（含扩展字段+K线）
export async function getStockDetail(code: string): Promise<StockDetail | null> {
  const prefix = code.startsWith('6') ? 'sh' : 'sz'
  const fullCode = `${prefix}${code}`

  // 获取实时详情
  const url = `${TENCENT_API}${fullCode}`
  const raw = await fetchTencentGBK(url)
  const detail = parseStockDetail(raw)
  if (!detail) return null

  // 尝试获取K线数据（最近60天）
  try {
    const today = new Date()
    const start = new Date(today.getTime() - 120 * 24 * 60 * 60 * 1000)
    const startStr = start.toISOString().slice(0, 10).replace(/-/g, '')
    const endStr = today.toISOString().slice(0, 10).replace(/-/g, '')
    const klineUrl = `https://web.ifzq.gtimg.cn/appstock/app/fqkline/get?param=${fullCode},day,${startStr},${endStr},120,qfq`
    const klineRaw = await axios.get(klineUrl, { timeout: 8000 })
    const klineData = parseKLineResponse(JSON.stringify(klineRaw.data), fullCode)
    detail.kline = klineData
  } catch {
    detail.kline = []
  }

  return detail
}

File: packages/backend-java/src/main/java/com/platform/backend/service/StockService.java
package com.platform.backend.service;

import com.platform.backend.model.*;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.Charset;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class StockService {

    private static final String TENCENT_API = "https://qt.gtimg.cn/q=";
    private static final String KLINE_API = "https://web.ifzq.gtimg.cn/appstock/app/fqkline/get";
    private static final String MINUTE_API = "https://web.ifzq.gtimg.cn/appstock/app/minute/query";
    private static final long CACHE_TTL = 5000;

    private final RestTemplate restTemplate = new RestTemplate();
    private final Map<String, CacheEntry<?>> cache = new ConcurrentHashMap<>();

    private record CacheEntry<T>(T data, long timestamp) {}

    private final List<StockInfo> stockList;

    public StockService() {
        List<StockInfo> loaded = new ArrayList<>();
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode root = mapper.readTree(getClass().getResourceAsStream("/stocks.json"));
            for (JsonNode node : root) {
                loaded.add(new StockInfo(
                    node.get("code").asText(),
                    node.get("name").asText(),
                    node.get("market").asText()
                ));
            }
        } catch (Exception e) {
            // Fallback to minimal list if file missing
            loaded.addAll(List.of(
                new StockInfo("sh600519", "贵州茅台", "上海"),
                new StockInfo("sz000001", "平安银行", "深圳"),
                new StockInfo("sz000002", "万科A", "深圳"),
                new StockInfo("sz300033", "同花顺", "深圳"),
                new StockInfo("sz000858", "五粮液", "深圳"),
                new StockInfo("sz002594", "比亚迪", "深圳")
            ));
        }
        this.stockList = List.copyOf(loaded);
    }

    private final List<StockInfo> indexList = List.of(
        new StockInfo("sh000001", "上证指数", ""),
        new StockInfo("sz399001", "深证成指", ""),
        new StockInfo("sz399006", "创业板指", ""),
        new StockInfo("sh000688", "科创50", "")
    );

    private final List<CommodityInfo> commodityList = List.of(
        new CommodityInfo("hf_GC", "COMEX黄金"),
        new CommodityInfo("hf_CL", "NYMEX原油"),
        new CommodityInfo("hf_HG", "COMEX铜"),
        new CommodityInfo("hf_SI", "COMEX白银"),
        new CommodityInfo("hf_NG", "NYMEX天然气")
    );

    public record StockInfo(String code, String name, String market) {}
    private record CommodityInfo(String code, String name) {}

    @SuppressWarnings("unchecked")
    private <T> T getCached(String key, java.util.function.Supplier<T> supplier) {
        CacheEntry<T> entry = (CacheEntry<T>) cache.get(key);
        if (entry != null && System.currentTimeMillis() - entry.timestamp < CACHE_TTL) {
            return entry.data;
        }
        T data = supplier.get();
        cache.put(key, new CacheEntry<>(data, System.currentTimeMillis()));
        return data;
    }

    private String fetchTencentGBK(String url) {
        ResponseEntity<byte[]> response = restTemplate.getForEntity(url, byte[].class);
        byte[] body = response.getBody();
        if (body == null) return "";
        return new String(body, Charset.forName("GBK"));
    }

    public List<StockQuote> getRealTimeQuotes(List<String> codes) {
        List<String> targetCodes = (codes != null && !codes.isEmpty()) ? codes : stockList.stream().map(StockInfo::code).toList();
        String cacheKey = String.join(",", targetCodes);
        return getCached(cacheKey, () -> {
            String url = TENCENT_API + String.join(",", targetCodes);
            String raw = fetchTencentGBK(url);
            return parseTencentResponse(raw);
        });
    }

    public List<StockInfo> getStockList() {
        return stockList;
    }

    public List<StockInfo> searchStocks(String keyword) {
        if (keyword == null || keyword.isBlank()) return List.of();
        String kw = keyword.toLowerCase();
        return stockList.stream()
            .filter(s -> s.code().toLowerCase().contains(kw) || s.name().contains(keyword))
            .toList();
    }

    public List<StockQuote> getMarketIndices() {
        return getCached("indices", () -> {
            String url = TENCENT_API + indexList.stream().map(StockInfo::code).collect(java.util.stream.Collectors.joining(","));
            String raw = fetchTencentGBK(url);
            return parseTencentResponse(raw);
        });
    }

    public List<CommodityQuote> getCommodities() {
        return getCached("commodities", () -> {
            String url = TENCENT_API + commodityList.stream().map(CommodityInfo::code).collect(java.util.stream.Collectors.joining(","));
            String raw = fetchTencentGBK(url);
            return parseCommodityResponse(raw);
        });
    }

    public StockDetail getStockDetail(String code) {
        String prefix = code.startsWith("6") ? "sh" : "sz";
        String fullCode = prefix + code;

        String url = TENCENT_API + fullCode;
        String raw = fetchTencentGBK(url);
        StockDetail detail = parseStockDetail(raw, code);
        if (detail == null) return null;

        // K线数据
        try {
            LocalDate today = LocalDate.now();
            LocalDate start = today.minusDays(120);
            String startStr = start.format(DateTimeFormatter.ofPattern("yyyyMMdd"));
            String endStr = today.format(DateTimeFormatter.ofPattern("yyyyMMdd"));
            String klineUrl = KLINE_API + "?param=" + fullCode + ",day," + startStr + "," + endStr + ",120,qfq";
            String klineRaw = restTemplate.getForObject(klineUrl, String.class);
            List<KLineData> kline = parseKLineResponse(klineRaw, fullCode);
            detail = new StockDetail(
                detail.code(), detail.name(), detail.price(), detail.change(), detail.changePercent(),
                detail.volume(), detail.amount(), detail.high(), detail.low(), detail.open(), detail.prevClose(),
                detail.market(), detail.updateTime(), detail.bidPrice(), detail.bidVolume(), detail.askPrice(), detail.askVolume(),
                detail.turnoverRate(), detail.peRatio(), detail.pbRatio(), detail.totalMarketCap(), detail.floatMarketCap(),
                detail.totalShares(), detail.floatShares(), detail.amplitude(), detail.volumeRatio(), detail.commissionRatio(),
                detail.avgPrice(), detail.week52High(), detail.week52Low(), detail.change5d(), detail.change10d(), detail.change20d(),
                kline
            );
        } catch (Exception e) {
            // ignore
        }
        return detail;
    }

    public List<MinuteData> getMinuteData(String code) {
        String prefix = code.startsWith("6") ? "sh" : "sz";
        String fullCode = prefix + code;
        String url = MINUTE_API + "?code=" + fullCode;
        try {
            String raw = restTemplate.getForObject(url, String.class);
            return parseMinuteResponse(raw, fullCode);
        } catch (Exception e) {
            return List.of();
        }
    }

    // ========== Parsers ==========

    private List<StockQuote> parseTencentResponse(String raw) {
        List<StockQuote> results = new ArrayList<>();
        String[] lines = raw.trim().split(";");
        for (String line : lines) {
            if (line.isBlank()) continue;
            int eq = line.indexOf("=\"");
            if (eq == -1) continue;
            int end = line.lastIndexOf('"');
            if (end <= eq) continue;
            String content = line.substring(eq + 2, end);
            String[] parts = content.split("~");
            if (parts.length < 45) continue;

            String name = parts[1];
            String code = parts[2];
            double price = parseDouble(parts[3]);
            double prevClose = parseDouble(parts[4]);
            double open = parseDouble(parts[5]);
            long volume = parseLong(parts[6]);
            double high = parseDouble(parts[33]);
            double low = parseDouble(parts[34]);
            double amount = parseDouble(parts[37]);
            String updateTime = parts[45];

            double change = price - prevClose;
            double changePercent = prevClose > 0 ? (change / prevClose) * 100 : 0;

            String market = code.startsWith("6") ? "上海" : "深圳";
            String stockName = name;
            // Only override if the API didn't return a name
            if (stockName == null || stockName.isBlank()) {
                for (StockInfo info : stockList) {
                    if (info.code.equals("sh" + code) || info.code.equals("sz" + code)) {
                        stockName = info.name;
                        market = info.market;
                        break;
                    }
                }
                if (stockName == null || stockName.isBlank()) {
                    for (StockInfo info : indexList) {
                        if (info.code.equals("sh" + code) || info.code.equals("sz" + code)) {
                            stockName = info.name;
                            market = "";
                            break;
                        }
                    }
                }
            }

            results.add(new StockQuote(
                code, stockName, price, round(change), round(changePercent),
                volume, amount, high, low, open, prevClose, market, updateTime
            ));
        }
        return results;
    }

    private List<CommodityQuote> parseCommodityResponse(String raw) {
        List<CommodityQuote> results = new ArrayList<>();
        String[] lines = raw.trim().split(";");
        for (String line : lines) {
            if (line.isBlank()) continue;
            int eq = line.indexOf("=");
            if (eq == -1) continue;
            String codeKey = line.substring(2, eq);
            int q1 = line.indexOf('"', eq);
            int q2 = line.lastIndexOf('"');
            if (q1 == -1 || q2 <= q1) continue;
            String content = line.substring(q1 + 1, q2);
            String[] parts = content.split(",");
            if (parts.length < 8) continue;

            double price = parseDouble(parts[0]);
            double changePercent = parseDouble(parts[1]);
            double open = parseDouble(parts[2]);
            double prevClose = parseDouble(parts[3]);
            double high = parseDouble(parts[4]);
            double low = parseDouble(parts[5]);
            String updateTime = parts[6];

            String name = codeKey;
            for (CommodityInfo info : commodityList) {
                if (info.code.equals(codeKey)) {
                    name = info.name;
                    break;
                }
            }

            results.add(new CommodityQuote(
                codeKey, name, round(price), round(changePercent), round(open),
                round(prevClose), round(high), round(low), updateTime
            ));
        }
        return results;
    }

    private StockDetail parseStockDetail(String raw, String inputCode) {
        String[] lines = raw.trim().split(";");
        if (lines.length == 0) return null;
        String line = lines[0];
        int eq = line.indexOf("=\"");
        if (eq == -1) return null;
        int end = line.lastIndexOf('"');
        if (end <= eq) return null;
        String content = line.substring(eq + 2, end);
        String[] p = content.split("~");
        if (p.length < 50) return null;

        String code = p[2];
        double price = parseDouble(p[3]);
        double prevClose = parseDouble(p[4]);
        double open = parseDouble(p[5]);
        long volume = parseLong(p[6]);
        double high = parseDouble(p[33]);
        double low = parseDouble(p[34]);
        double amount = parseDouble(p[37]);

        List<Double> bidPrice = new ArrayList<>();
        List<Integer> bidVolume = new ArrayList<>();
        List<Double> askPrice = new ArrayList<>();
        List<Integer> askVolume = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            bidPrice.add(parseDouble(p[9 + i * 2]));
            bidVolume.add((int) parseLong(p[10 + i * 2]));
            askPrice.add(parseDouble(p[19 + i * 2]));
            askVolume.add((int) parseLong(p[20 + i * 2]));
        }

        double change = price - prevClose;
        double changePercent = prevClose > 0 ? (change / prevClose) * 100 : 0;
        String market = code.startsWith("6") ? "上海" : "深圳";
        String name = p[1];
        if (name == null || name.isBlank()) {
            for (StockInfo info : stockList) {
                if (info.code.equals("sh" + code) || info.code.equals("sz" + code)) {
                    name = info.name;
                    market = info.market;
                    break;
                }
            }
            if (name == null || name.isBlank()) {
                for (StockInfo info : indexList) {
                    if (info.code.equals("sh" + code) || info.code.equals("sz" + code)) {
                        name = info.name;
                        market = "";
                        break;
                    }
                }
            }
        }

        return new StockDetail(
            code, name, price, round(change), round(changePercent), volume, amount, high, low, open, prevClose, market, p[30] != null ? p[30] : "",
            bidPrice, bidVolume, askPrice, askVolume,
            parseDouble(p[38]), parseDouble(p[39]), parseDouble(p[46]),
            parseDouble(p[44]), parseDouble(p[45]), parseDouble(p[67]), parseDouble(p[68]),
            parseDouble(p[43]), parseDouble(p[49]), parseDouble(p[50]), parseDouble(p[51]),
            parseDouble(p[47]), parseDouble(p[48]), parseDouble(p[62]), parseDouble(p[63]), parseDouble(p[64]),
            List.of()
        );
    }

    @SuppressWarnings("unchecked")
    private List<KLineData> parseKLineResponse(String raw, String stockCode) {
        if (raw == null) return List.of();
        try {
            // Simple JSON parsing for the specific Tencent format
            int dataIdx = raw.indexOf("\"data\":");
            if (dataIdx == -1) return List.of();
            int codeIdx = raw.indexOf("\"" + stockCode + "\":");
            if (codeIdx == -1) return List.of();
            int dayIdx = raw.indexOf("\"day\":", codeIdx);
            if (dayIdx == -1) return List.of();
            int arrStart = raw.indexOf("[", dayIdx);
            int arrEnd = raw.indexOf("]", arrStart);
            if (arrStart == -1 || arrEnd == -1) return List.of();
            String arrContent = raw.substring(arrStart + 1, arrEnd);
            List<KLineData> result = new ArrayList<>();
            // Split by "],[" to get each day's data
            String[] days = arrContent.split("\\],\\[");
            for (String day : days) {
                day = day.replace("[", "").replace("]", "").replace("\"", "");
                String[] parts = day.split(",");
                if (parts.length < 6) continue;
                result.add(new KLineData(
                    parts[0].trim(),
                    parseDouble(parts[1].trim()),
                    parseDouble(parts[2].trim()),
                    parseDouble(parts[3].trim()),
                    parseDouble(parts[4].trim()),
                    parseLong(parts[5].trim()),
                    parts.length > 6 ? parseDouble(parts[6].trim()) : 0
                ));
            }
            return result;
        } catch (Exception e) {
            return List.of();
        }
    }

    @SuppressWarnings("unchecked")
    private List<MinuteData> parseMinuteResponse(String raw, String stockCode) {
        if (raw == null) return List.of();
        try {
            int codeIdx = raw.indexOf("\"" + stockCode + "\":");
            if (codeIdx == -1) return List.of();
            int dataIdx = raw.indexOf("\"data\":", codeIdx);
            if (dataIdx == -1) return List.of();
            int innerData = raw.indexOf("\"data\":", dataIdx + 7);
            if (innerData == -1) return List.of();
            int arrStart = raw.indexOf("[", innerData);
            int arrEnd = raw.indexOf("]", arrStart);
            if (arrStart == -1 || arrEnd == -1) return List.of();
            String arrContent = raw.substring(arrStart + 1, arrEnd);
            List<MinuteData> result = new ArrayList<>();
            String[] items = arrContent.split(",");
            for (String item : items) {
                item = item.replace("\"", "").trim();
                String[] parts = item.split(" ");
                if (parts.length < 4) continue;
                result.add(new MinuteData(
                    parts[0], parseDouble(parts[1]), parseLong(parts[2]), parseDouble(parts[3])
                ));
            }
            return result;
        } catch (Exception e) {
            return List.of();
        }
    }

    private double parseDouble(String s) {
        try { return Double.parseDouble(s); } catch (Exception e) { return 0; }
    }

    private long parseLong(String s) {
        try { return Long.parseLong(s); } catch (Exception e) { return 0; }
    }

    private double round(double n) {
        return Math.round(n * 100) / 100.0;
    }
}
